File: facade/src/main/java/org/tdf/sunflower/net/PeerServer.java
package org.tdf.sunflower.net;

import org.tdf.sunflower.facade.PeerServerListener;

import java.util.Collections;
import java.util.List;
import java.util.Properties;

public interface PeerServer {
    PeerServer NONE = new PeerServer() {
        @Override
        public void dial(Peer peer, byte[] message) {

        }

        @Override
        public void broadcast(byte[] message) {

        }

        @Override
        public List<Peer> getPeers() {
            return Collections.emptyList();
        }

        @Override
        public List<Peer> getBootStraps() {
            return Collections.emptyList();
        }

        @Override
        public void addListeners(PeerServerListener... peerServerListeners) {

        }

        @Override
        public void start() {

        }

        @Override
        public void stop() {

        }

        @Override
        public void init(Properties properties) {

        }

        @Override
        public Peer getSelf() {
            return Peer.NONE;
        }

        @Override
        public boolean isFull() {
            return false;
        }
    };

    // dial a peer with a message
    void dial(Peer peer, byte[] message);

    // broadcast a message to all the peers
    void broadcast(byte[] message);

    // get all peers had been connected
    List<Peer> getPeers();

    // get all bootstraps
    List<Peer> getBootStraps();

    boolean isFull();

    void addListeners(PeerServerListener... peerServerListeners);

    void start();

    void stop();

    void init(Properties properties);

    Peer getSelf();
}

File: facade/src/main/java/org/tdf/sunflower/net/PeerServer.kt
package org.tdf.sunflower.net

import org.tdf.sunflower.facade.PeerServerListener

interface PeerServer {
    // dial a peer with a message
    fun dial(peer: Peer, message: ByteArray)

    // broadcast a message to all the peers
    fun broadcast(message: ByteArray)

    // get all peers had been connected
    val peers: List<Peer>

    // get all bootstraps
    val bootstraps: List<Peer>
    val isFull: Boolean
    fun addListeners(vararg peerServerListeners: PeerServerListener)
    fun start()
    fun stop()
    val self: Peer

    companion object {
        @JvmField
        val NONE: PeerServer = object : PeerServer {
            override fun dial(peer: Peer, message: ByteArray) {}
            override fun broadcast(message: ByteArray) {}
            override val peers: List<Peer>
                get() = emptyList()
            override val bootstraps: List<Peer>
                get() = emptyList()

            override fun addListeners(vararg peerServerListeners: PeerServerListener) {}
            override fun start() {}
            override fun stop() {}
            override val self: Peer
                get() = Peer.NONE
            override val isFull: Boolean
                get() = false
        }
    }
}